### PR TITLE
Sqrtmod

### PR DESCRIPTION
**Pull request by abh2k**
* ntheory
  *  In `def _sqrt_mod_prime_power(a, p, k):` it is assumed that p is a prime number. And this function calls `if not is_quad_residue(a, p):` in line 396. But `is_quad_residue(a, p)` again checks that p is prime or not. This leads to extra computation cost.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
The tests have failed. Maybe it's necessary to check whether p is odd before calling `legendre_symbol`.

**Comment by oscarbenjamin** [comment]
Can you edit the OP to explain what the purpose of this is?

The release note is not really appropriate: it should have relevant information for users. If this is just intended to speed up the function then that's all it should say in the release note.

**Comment by abhinav28071999** [comment]
> The tests have failed. Maybe it's necessary to check whether p is odd before calling `legendre_symbol`.

Yes you are right. I will add  p != 2.

**Comment by jksuom** [comment]
Can you also add some data on what kind of speed-up is to be expected.

**Comment by abhinav28071999** [comment]
> Can you also add some data on what kind of speed-up is to be expected.

I think that this will not make any change. While going through the code i found that this is already implemented in ''def _sqrt_mod_prime_power(a, p, k):'' function at line 396.
`        if not is_quad_residue(a, p):
            return None` . 
But i noticed that this is inefficient, because in  `def _sqrt_mod_prime_power(a, p, k):` we assumed that p is a prime number and but the function `is_quad_residue` implements isprime() which is not necessary

File: sympy/ntheory/residue_ntheory.py [hunk 1]
```diff
@@ -393,7 +393,7 @@ def _sqrt_mod_prime_power(a, p, k):
     if k == 1:
         if p == 2:
             return [ZZ(a)]
-        if not is_quad_residue(a, p):
+        if not (pow(a, (p - 1) // 2, p) == 1 or p < 3 or a % p < 2):
```
Comment (jksuom) [inline sympy/ntheory/residue_ntheory.py]: `p < 3` means that `p == 2` that was already tested above.

Comment (abhinav28071999) [reply sympy/ntheory/residue_ntheory.py]: yes, i will change it.

Comment (jksuom) [reply sympy/ntheory/residue_ntheory.py]: I would also consider taking the quick test `a % p < 2` first before the `pow()` condition.

Comment (abhinav28071999) [reply sympy/ntheory/residue_ntheory.py]: Done.

**Comment by jksuom** [comment]
Thanks. Looks good.

**Pull request merged**